proxy/apifilterproxy/filters/auth/rancher_token_validate_filter.go

```diff
@@ -0,0 +1,277 @@
+		for i := range cookie {
+			if strings.Contains(cookie[i], "token") {
+				cookieString = cookie[i]
+			}
+		}
+	} else {
+		log.Debugf("No token found in cookie.")
+		output.Status = http.StatusOK
+		return output, nil
+	}
+
+	tokens := strings.Split(cookieString, ";")
+	tokenValue := ""
+	if len(tokens) >= 1 {
+		for i := range tokens {
+			if strings.Contains(tokens[i], "token") {
+				if len(strings.Split(tokens[i], "=")) > 1 {
+					tokenValue = strings.Split(tokens[i], "=")[1]
+				}
+			}
+
+		}
+	} else {
+		log.Errorf("No token found")
+		output.Status = http.StatusOK
+		return output, nil
+	}
+	if tokenValue == "" {
+		log.Errorf("No token found")
+		output.Status = http.StatusOK
+		return output, nil
+	}
+
+	//check if the token value is empty or not
+	if tokenValue != "" {
+		log.Debugf("token:" + tokenValue)
+		log.Debugf("envid:" + envid)
+		projectID, accountID := "", ""
+		var err error
+		if envid != "" {
+			projectID, accountID, err = getAccountAndProject(f.rancherURL, envid, tokenValue)
+			if accountID == "Unauthorized" {
+				log.Errorf("Unauthorized")
+				output.Status = http.StatusUnauthorized
+				return output, nil
+			}
+
+			if accountID == "Forbidden" {
+				log.Errorf("Forbidden")
+				output.Status = http.StatusForbidden
+				return output, nil
+			}
+			if err != nil {
+				log.Errorf("Error getting the accountid and projectid: %v", err)
+				output.Status = http.StatusNotFound
+				return output, nil
+			}
+		} else {
+			accountID, err = getAccountID(f.rancherURL, tokenValue)
+			if accountID == "Unauthorized" {
+				log.Errorf("Unauthorized")
+				output.Status = http.StatusUnauthorized
+				return output, nil
+			}
+			if err != nil {
+				log.Errorf("Error getting the accountid : %v", err)
+				output.Status = http.StatusNotFound
+				return output, nil
+			}
+		}
+
+		//construct the responseBody
+		var headerBody = make(map[string][]string)
+
+		requestHeader := input.Headers
+		for k, v := range requestHeader {
+			headerBody[k] = v
+		}
+
+		headerBody["X-API-Account-Id"] = []string{accountID}
+		if projectID != "" {
+			headerBody["X-API-Project-Id"] = []string{projectID}
+		}
+
+		output.Headers = headerBody
+		output.Status = http.StatusOK
+
+		log.Debugf("Response <= %v", output)
+	}
+	return output, nil
+}
+
+//get the projectID and accountID from rancher API
+func getAccountAndProject(host string, envid string, token string) (string, string, error) {
+
+	client := &http.Client{}
+	requestURL := host + "/v2-beta/projects/" + envid + "/accounts"
+	log.Debugf("requestURL %v", requestURL)
+	req, err := http.NewRequest("GET", requestURL, nil)
+	if err != nil {
+		log.Errorf("Cannot connect to the rancher server. Please check the rancher server URL")
+		return "", "", err
+	}
+	cookie := http.Cookie{Name: "token", Value: token}
+	req.AddCookie(&cookie)
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Errorf("Cannot connect to the rancher server. Please check the rancher server URL")
```

Comment:
we already log the error in the caller stack. There is no need to log it here. What you can do is - return the error in the format:

```
return ("", "", fmt.Errorf("Something failed due to [%v]", err))
```

Same for all the code blocks below